### PR TITLE
Add bucket id calculation function

### LINKED ISSUE #76: Add bucket_id calculation function
After #71 we'll have all information that is necessary for calculating bucket_id by a sharding key. The proposal is to implement this helper function in the module.

## Example

Schema:

```
TBD
```

Assume we serve HTTP requests:

```lua
function handler(req)
    local bucket_id = ddl.bucket_id('users', req:json()['user_id'])
    local ok, err = vshard.router.callrw(bucket_id, <..args..>)
    <...>
end
```

The key point is that ddl knows, which sharding function is is use. Without the function we would do the following:

```lua
function handler(req)
    local bucket_id = vshard.router.bucket_id_mpcrc32(req:json()['user_id']) -- !!
    local ok, err = vshard.router.callrw(bucket_id, <..args..>)
    <...>
end
```

### LINKED ISSUE #83: Wrong sharding function check?
https://github.com/tarantool/ddl/blob/c88c3eacb530a4739311819b590953735833b957/ddl/check.lua#L745-L764

Say, for 'foo.bar' we'll check `_G['foo.bar']`, not `_G.foo.bar`.

~BTW, shouldn't we use strict-mode-checked access instead of `rawget(_G, <...>)`?~ (not within this issue).

### PR DESCRIPTION
**Pull request by ligurio**
Closes #76
Fixes #83

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 
 - Name and format of a space `_ddl_sharding_key` is a part of public API.
+- Allow specifying `sharding_func` for space.
```
Comment (rosik) [inline CHANGELOG.md:14]: Move it to unreleased section, please

File: README.md [hunk 2]
```diff
@@ -33,10 +33,39 @@ in [documementation](https://www.tarantool.io/en/doc/latest/book/box/data_model/
...
     Return values: `true` if no error, otherwise return `nil, err`
 
-Additionally when a section `sharding_key` is not empty in YAML schema call of
-function `ddl.set_schema(schema)` creates a space `_ddl_sharding_key` with two
+Call of function `ddl.set_schema(schema)` creates a space `_ddl_sharding_key` with two
 fields: `space_name` with type `string` and `sharding_key` with type `array`.
 
+Similarly for `sharding_func`: call of function `ddl.set_schema(schema)` creates
+a space `_ddl_sharding_func` with three fields: `space_name` with type `string`,
+`sharding_func_name` with type `string` and `sharding_func_body` with type `string`.
```
Comment (rosik) [inline README.md:41]: Are sharding_func_name and sharding_func_body are mutually exclusive?

**Review by rosik — COMMENTED** [review]
Why do you create a new space for sharding functions? Why not enhance the `functions` section?

image

How do I define the same function for multiple spaces? Do I have to copy-paste the same body 10 times?

**Comment by ligurio** [comment]
@rosik Please stop review commit related to sharding function support. This PR is about buket_id() method.
Changes (review and conversation) for sharding function support are in https://github.com/tarantool/ddl/pull/72.

File: CHANGELOG.md (hunk 1 above)
Comment (AnaNek) [reply CHANGELOG.md:14]: Updated:

```
 ## [Unreleased]

 ### Added
 
 - Added integration with service coveralls.io.
+- Allow specifying `sharding_func` for space.
 
 ## [1.5.0] - 2021-08-09
 
 ### Added
 
 - Name and format of a space `_ddl_sharding_key` is a part of public API.
-- Allow specifying `sharding_func` for space.
```

File: README.md (hunk 2 above)
Comment (AnaNek) [reply README.md:41]: Yes, if user specified sharding function name then information about sharding func is stored in `sharding_func_name` field and `sharding_func_body` is null. If user specified sharding func as table like `{ body = 'function(key) return <...> end' }` then information is stored in `sharding_func_body` field and `sharding_func_name` field is null.

File: README.md [hunk 3]
```diff
@@ -33,10 +33,39 @@ in [documementation](https://www.tarantool.io/en/doc/latest/book/box/data_model/
...
+Also, you can pass your own sharding function by defining the function name in `_G`
+or by specifying lua code directly in `body` field:
+```lua
+sharding_func = {
+  body = 'function(key) return <...> end'
+}
+```
+
+Your defined sharding function in `_G` should have type `function`
+or type `table` with `__call` method.
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply README.md:57]: Why? We have case when user specified sharding function as Lua code (table with `body` field) in schema and case when user specified sharding function name and this name refers to an object in a global environment `_G`. User can specify this object by himself in `_G`.

File: ddl.lua [hunk 4]
```diff
@@ -163,8 +163,37 @@ local function get_schema()
...
+local function bucket_id(space_name, sharding_key)
+    if type(space_name) ~= 'string' then
+        return nil, string.format(
+            "Invalid space name (string expected, got %s)", type(space_name)
+        )
+    end
+    if sharding_key == nil then
+        return nil, string.format(
+            "Sharding key specified for space (%s) is nil", space_name)
+    end
```
Comment (Totktonada) [inline ddl.lua:175]: It is covered by the if statement above:

```lua
tarantool> type(nil)
---
- nil
...

tarantool> function foo(x) return type(x) end
tarantool> foo(nil)
---
- nil
...

tarantool> foo()
---
- nil
...
```

File: ddl/get.lua [hunk 5]
```diff
@@ -111,6 +111,62 @@ local function get_space_schema(space_name)
...
+    end
+
+    if type(sharding_func) == 'string' then
+        if sharding_func:find('^vshard.') ~= nil then
+            sharding_func = string.format(vshard_func_body, sharding_func)
+        else
+            sharding_func = 'return ' .. sharding_func
+        end
+        return assert(loadstring(sharding_func))()
+    end
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply ddl/get.lua]: 1. Here we add exactly same problem for `ddl.bucket_id()` as described in #83 for `ddl.check_schema()`: `{sharding_func = 'foo.bar'}` must mean `_G.foo.bar` (not `_G['foo.bar']`). It seems, we cannot implement this code correctly if we'll not fix #83 first. I'm against of introducting the bug here and resolve it later: this way we'll give some kind of indulgence for the wrong behaviour. It'll be harder to understand what is right and what is wrong, if we'll look over this commit in a future.
2. Nit: I would check whether the type is `string` and only in this case try to access the `_G` field.

File: ddl/get.lua [hunk 6]
```diff
@@ -111,6 +111,62 @@ local function get_space_schema(space_name)
...
+
+    local vshard_func_body = [[ return function(key)
+        local vshard = require('vshard')
+        return %s(key)
+    end
+    ]]
+
+    if type(sharding_func) == 'function' then
+        return sharding_func
+    end
```
[the remark this answers is not in the document]
Comment (Totktonada) [reply ddl/get.lua]: (Not a callable object itself, but it's name.) Okay.

File: test/bucket_id_test.lua [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+    -- Create a userdata object.
+    local userdata_obj = newproxy(true)
+    local mt = getmetatable(userdata_obj)
+    mt.__call = helper.sharding_func
+    rawset(_G, sharding_func_name, userdata_obj)
+    t.assert_equals(type(userdata_obj), 'userdata')
+
+    -- Set sharding function as a __call method of userdata object.
+    rawset(_G, sharding_func_name, userdata_obj)
```
Comment (Totktonada) [inline test/bucket_id_test.lua:172]: 1. The comment does not look relevant to the code.
2. The same rawset() is four lines above.

File: test/bucket_id_test.lua [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	    double y;
+       } point_t;
+    ]]
+    local point = ffi.metatype('point_t', {
+        __call = helper.sharding_func
+    })
+    t.assert_equals(type(point), 'cdata')
+
+    -- Set sharding function as a __call method of cdata object.
+    rawset(_G, sharding_func_name, point())
```
Comment (Totktonada) [inline test/bucket_id_test.lua:203]: 1. The comment does not look relevant to the code (we don't set a metatype with __call here).
2. Strange. The cdata object does not participate in the test at all (only `point()` result, but `point` should be here). I don't know what is going on here, but it seems that the test should fail on `ddl.set_schema()`.

File: test/bucket_id_test.lua (hunk 7 above)
Comment (ligurio) [reply test/bucket_id_test.lua:172]: Fixed 1 and 2:
```lua
--- a/test/bucket_id_test.lua
+++ b/test/bucket_id_test.lua
@@ -161,14 +161,13 @@ end
 function g.test_bucket_id_sharding_func_userdata_call()
     local sharding_func_name = 'sharding_func_userdata_call'
 
-    -- Create a userdata object.
+    -- Create a userdata object and set sharding function as a __call method of
+    -- userdata object.
     local userdata_obj = newproxy(true)
     local mt = getmetatable(userdata_obj)
     mt.__call = helper.sharding_func
-    rawset(_G, sharding_func_name, userdata_obj)
     t.assert_equals(type(userdata_obj), 'userdata')
 
-    -- Set sharding function as a __call method of userdata object.
     rawset(_G, sharding_func_name, userdata_obj)
     g.schema.spaces.space.sharding_func = sharding_func_name
```

File: test/bucket_id_test.lua (hunk 8 above)
Comment (ligurio) [reply test/bucket_id_test.lua]: 1. Fixed.
2. Indeed test for cdata was incorrect.

```lua
--- a/test/bucket_id_test.lua
+++ b/test/bucket_id_test.lua
@@ -200,12 +200,13 @@ function g.test_bucket_id_sharding_func_cdata_call()
            double y;
        } point_t;
     ]]
-    local point = ffi.metatype('point_t', {
+    ffi.metatype('point_t', {
         __call = helper.sharding_func
     })
+    local point = ffi.new('point_t')
     t.assert_equals(type(point), 'cdata')
 
-    rawset(_G, sharding_func_name, point())
+    rawset(_G, sharding_func_name, point)
     g.schema.spaces.space.sharding_func = sharding_func_name
 
     -- Set DDL schema.
```

File: ddl/get.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply ddl/get.lua]: Fixed both issues.
Fix for #83 added in a separate commit.

File: test/set_sharding_metadata_test.lua [hunk 9]
```diff
@@ -420,10 +424,11 @@ function g.test_ddl_user_sharding_func_cdata_call()
         } test_set_struct_t;
     ]]
 
-    local test_set_struct = ffi.metatype('test_set_struct_t', {
+    ffi.metatype('test_set_struct_t', {
         __call = function(_, key) return key end
     })
-    rawset(_G, user_sharding_func_name, test_set_struct())
+    local test_set_struct = ffi.new('test_set_struct_t')
+    rawset(_G, user_sharding_func_name, test_set_struct)
```
Comment (AnaNek) [inline test/set_sharding_metadata_test.lua:431]: Can you please highlight for me what is wrong with this test and what is the difference?

Old version:

```lua
tarantool> ffi = require('ffi')
---
...

tarantool> ffi.cdef[[
        typedef struct
        {
            int data;
        } test_set_struct_t;
    ]]
---
...

tarantool> test_set_struct = ffi.metatype('test_set_struct_t', {
        __call = function(_, key) return key end
    })
---
...

tarantool> user_sharding_func = test_set_struct()
---
...

tarantool> user_sharding_func(2)
---
- 2
...

```

New version:

```lua
tarantool> ffi = require('ffi')
---
...

tarantool> ffi.cdef[[
        typedef struct
        {
            int data;
        } test_set_struct_t;
    ]]
---
...

tarantool> ffi.metatype('test_set_struct_t', {
         __call = function(_, key) return key end
     })
---
- ctype<struct 1071>
...

tarantool> test_set_struct = ffi.new('test_set_struct_t')
---
...

tarantool> user_sharding_func = test_set_struct
---
...

tarantool> user_sharding_func(2)
---
- 2
...
```

Comment (ligurio) [reply test/set_sharding_metadata_test.lua:431]: > Can you please highlight for me what is wrong with this test and what is the difference?

Nothing wrong, actually. I found a new version of assigning metatable to cdata more clear and decide to rewrite similar code in your tests too. If it makes you sad, I can revert a patch.

Comment (Totktonada) [reply test/set_sharding_metadata_test.lua:431]: > Nothing wrong, actually.

But your commit message tells us: 'Fix tests for <...>'. Something that is not broken cannot be fixed. Refactored, cleaned up, clarified, simplified, revisited, transformed, rewritten — okay.

File: test/bucket_id_test.lua [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+
+function g.test_bucket_id_sharding_func_cdata_call()
+    local sharding_func_name = 'sharding_func_cdata_call'
+
+    -- Create a cdata object and set sharding function as a __call method of
+    -- cdata object.
+    ffi.cdef[[
+       typedef struct {
+	    double x;
+	    double y;
```
Comment (Totktonada) [inline test/bucket_id_test.lua:215]: Tabs instead of spaces (and below too).

File: ddl/get.lua [hunk 11]
```diff
@@ -111,6 +114,56 @@ local function get_space_schema(space_name)
...
+            return sharding_func
+        end
+    end
+
+    if type(sharding_func_def) == 'table' then
+        local sharding_func = loadstring('return ' .. sharding_func_def.body)
+        if sharding_func == nil then
+            return nil, string.format(
+                "Body is incorrect in sharding_func for space (%s)", space_name)
+        end
```
Comment (Totktonada) [inline ddl/get.lua:131]: If something goes wrong, `loadstring()` returns an error in the second return value. It worth to concatenate it into our error.

**Review by Totktonada — APPROVED** [review]
LGTM aside a few noted points.

File: test/bucket_id_test.lua (hunk 10 above)
Comment (ligurio) [reply test/bucket_id_test.lua:215]: Fixed in both places.

File: ddl/get.lua (hunk 11 above)
Comment (ligurio) [reply ddl/get.lua:131]: Added err to message:
```diff
--- a/ddl/get.lua
+++ b/ddl/get.lua
@@ -124,10 +124,10 @@ local function prepare_sharding_func_for_call(space_name, sharding_func_def)
     end
 
     if type(sharding_func_def) == 'table' then
-        local sharding_func = loadstring('return ' .. sharding_func_def.body)
+        local sharding_func, err = loadstring('return ' .. sharding_func_def.body)
         if sharding_func == nil then
             return nil, string.format(
-                "Body is incorrect in sharding_func for space (%s)", space_name)
+                "Body is incorrect in sharding_func for space (%s): %s", space_name, err)
         end
         return sharding_func()
     end
```

File: test/set_sharding_metadata_test.lua (hunk 9 above)
Comment (ligurio) [reply test/set_sharding_metadata_test.lua:431]: Updated commit message and moved commit on top of the branch to order changes in a logical way.
